### PR TITLE
Reimplement server resource

### PR DESCRIPTION
**Pull request by nvthongswansea**

### CODE REVIEW

File: gridscale/relation-manager/serverRelationManager.go [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+//getData returns resource data from server relation manager
+func (c ServerRelationManger) getData() *schema.ResourceData {
+	return c.data
+}
+
+//LinkStorages links storages to a server
+//**Note: if no storage is set as the boot device. The storage in the
+//1st order of the list will be set as the boot device automatically
```
Comment (bkircher) [inline gridscale/relation-manager/serverRelationManager.go]: Consider rewording:
- "The storage in the 1st order of the list will be set as the boot device automatically" to 
+ "The first storage in the list will be automatically set as the boot device"

File: gridscale/resource_gridscale_server.go [hunk 2]
```diff
@@ -68,9 +70,11 @@ func resourceGridscaleServer() *schema.Resource {
 				},
 			},
 			"storage": {
-				Type:     schema.TypeSet,
+				Type:     schema.TypeList,
 				Optional: true,
 				MaxItems: 8,
+				Description: `A list of storages attached to the server. The storage in the 1st order of the list is always the boot storage.
+In order to replace the boot storage, the new boot storage must be placed in the 1st order of the list.`,
 				Elem: &schema.Resource{
```
Comment (bkircher) [inline gridscale/resource_gridscale_server.go:77]: Consider rewording comment
- "The storage in the 1st order of the list is always the boot storage."
+ "The first storage in the list is always set as the boot storage of the server."

**Review by bkircher — CHANGES_REQUESTED** [review]
Okay, this is a big PR.
Reviewed up to commit dad2701.
First: Looks good! Awesome work. Thanks for this, Tom.

I think with this change it will be finally deterministic what storage is the actual boot storage (system disk) when creating new servers with multiple storages attached. IIRC, before you had to cross your fingers that the storage order was as you wanted it to be when executing Terraform. Am I right with this?

9af5ff6: I like how `resourceGridscaleServerCreate` implementation gets simpler with the `ServerRelationManger`.
1123fab: same thing, I like how `resourceGridscaleServerUpdate` gets simpler

Future suggestion:
In general, it would help in future when you could give the rationale for a change inside a commit message. (With rationale meaning the reason for the change, the "why".) This would make it easier to understand the context for a change. Take 580c9da for example:

```txt
commit 580c9da5e2afe62e5fb93b5738d18cbe051583fc
Author: nvthongswansea <nvthongswansea@gmail.com>
Date:   Thu Oct 24 13:52:26 2019 +0200

    add server dependency manager

    Server dependency manager features:
    - Create/Update/Remove server's realtions.
```

The commit message says what this commits adds `ServerRelationManger` (good) and it says also what `ServerRelationManger` is supposed to do (also good) but it misses to say why you think this is needed, what problem it solves. Makes reviewing easier ;-)

**Comment by nvthongswansea** [comment]
@bkircher thanks! And you are right about ordering storage 👍

**Review by bkircher — APPROVED** [review]
LGTM

**Pull request merged**